SSR-3683: Develop tooltip and fieldGroup Customization tips (#1495)
kintone-labs/kintone-ui-component: docs/document/i18n/ja/docusaurus-plugin-content-docs/current/guides/in-office-day-list-customization.md

### 使用するコンポーネント

- [Dropdown](../components/desktop/dropdown.md)
- [FieldGroup](../components/desktop/field-group.md)
- [MultiChoice](../components/desktop/multichoice.md)
- [Table](../components/desktop/table.md)
- [Text](../components/desktop/text.md)
- [Tooltip](../components/desktop/tooltip.md)

## 事前準備

以下のフィールドを含むアプリを作成します。

- スペースフィールド（要素 ID：space）

## JavaScript/CSS カスタマイズ

kintone UI Component の UMD ファイルをアプリに読み込んだ上で、以下のような実装をした JavaScript ファイルをアップロードします。<br/>
ファイルのアップロード方法などは、 [Quick Start](../getting-started/quick-start.md) をご覧ください。

### FieldGroup と Tooltip の基本的な使い方

#### FieldGroup 内で Tooltip を利用する

"KUC team" というラベルの FieldGroup を作成し、Text コンポーネントを追加します。

```javascript
const fieldGroup = new Kuc.FieldGroup({
  label: "KUC team",
  content: new Kuc.Text({ text: "This is a FieldGroup" }),
});
```

FieldGroup に "Office day info" と表示する Tooltip を追加します。

```javascript
const tooltip = new Kuc.Tooltip({
  title: "Office day info",
  content: fieldGroup,
});
```

#### UI 上の表示

![render](/img/tooltip_field-group.gif)

### FieldGroup に Table コンポーネントを追加する

kintone の標準機能では FieldGroup 内にテーブルを持つことはできませんが、kintone UI Component を使うことで FieldGroup を開いた際にテーブルを表示することができます。

> #### HTML Button に Tooltip を作成

Reviewer: ```suggestion
#### HTML Button に Tooltip を追加
```

> HTML ボタンに対して Tooltip コンポーネント追加し、"Submit changes or additions for this entry." (このエントリーの変更または追加を送信してください) というメッセージを出します。

Reviewer: ```suggestion
HTML ボタンに Tooltip コンポーネントを追加し、"Submit changes or additions for this entry." (この入力の変更または追加してください) というメッセージを出します。
```

```javascript
const button = document.createElement("button");
button.textContent = "Submit";

const tooltipForButton = new Kuc.Tooltip({
  title: "Submit changes or additions for this entry.",
  container: button,
  describeChild: true,
  placement: "bottom",
});
```

> #### Table にDropdown、Text、MultiChoice のコンポーネントと Tooltip が追加されたボタンを配置

Reviewer: ```suggestion
#### Table に Dropdown, Text, MultiChoice のコンポーネントと Tooltip が追加されたボタンを配置
```

      field: "day",
      render: renderDay,
    },
    {
      title: "Location",
      field: "location",
      render: renderLocation,
    },
    {
      title: "Note",
      field: "note",
      render: renderNote,
    },
    {
      title: "Action",
      field: "action",
      render: renderAction,
    },
  ],
  data: [
    {
      name: "",
      day: [],
      location: "-",
      note: "",
    },
  ],
});
```

#### FieldGroup に Table コンポーネントを追加

"KUC team" というラベルの FieldGroup を作成し、Table コンポーネントを追加します。

```javascript
const kucTeamFieldGroup = new Kuc.FieldGroup({
  label: "KUC team",
  content: kucTeamTable,
});
```

#### FieldGroup に Tooltip を追加

FieldGroup に Tooltip を追加します。

```javascript
const kucTeamTooltip = new Kuc.Tooltip({
  title: "Office day info",
  container: kucTeamFieldGroup,
});
```

#### FieldGroup にイベントリスナーを追加

FieldGroup の開閉で Tooltip のメッセージを変えるため、FieldGroup にイベントリスナーを追加します。

```javascript
kucTeamFieldGroup.addEventListener("change", (e) => {
  if (e.detail.expanded) {
    kucTeamFieldGroup.title = "";
  } else {
    kucTeamFieldGroup.title = "Office day info";
  }
});
```

#### UI 上の表示

![render](/img/tooltip_fieldgroup_customize.gif)

> ### Tooltip の２種類のユースケース

Reviewer: Please unify with プロパティ not プロパティー.

> ここでは [`describeChild`](../components/desktop/tooltip.md#property) というプロパティーを紹介します。

Reviewer: No need of `` for Japanese, because users can distinguish property name and normal sentence with JA and EN.
```suggestion
ここでは [describeChild](../components/desktop/tooltip.md#property) というプロパティーを紹介します。
```

このプロパティーはデフォルトでは false に設定されています。<br/>
describeChild プロパティの使い方を理解することで、アプリケーションのアクセシビリティとユーザビリティを効果的に向上させることができます。<br/>

> [screen reader](https://developer.mozilla.org/en-US/docs/Glossary/Screen_reader) を有効にして、ボタンへのフォーカスを動かしスクリーンリーダーの読みとりを確認してみましょう。<br/>

Reviewer: We have example for Button and Text, so I removed the mention of Button. Updated EN, too.

```suggestion
[スクリーンリーダー](https://developer.mozilla.org/ja/docs/Glossary/Screen_reader) を有効にして、フォーカスを動かしてスクリーンリーダーの読み取りを確認してみましょう。<br/>
```

> #### Tooltip represents the description of the element

Reviewer: ```suggestion
#### Tooltip が要素の説明を表す場合
```

Tooltip が要素の説明を表す（要素についての追加情報や補足説明を提供する）場合は、describeChild を true に設定します。

```javascript
const tooltipForButton = new Kuc.Tooltip({
  title: "Submit changes or additions for this entry.",
  container: button,
  describeChild: true,
  placement: "bottom",
});
```

![render](/img/tooltip_describeChild_true.gif)

> この例では、describeChild が true の場合、ボタンがフォーカスを得たときにスクリーンリーダーはまずボタン自体のコンテント "Submit" を読み、続けて Tooltip のタイトル "Submit changes or additions for this entry." を読み取ります。

Reviewer: ```suggestion
この例では、describeChild が true の場合、ボタンがフォーカスを得たときにスクリーンリーダーはまずボタン自体のコンテンツ "Submit" を読み、続けて Tooltip のタイトル "Submit changes or additions for this entry." を読み取ります。
```

> これはdescribeChild が true にされている場合、Tooltip は [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) を子要素(ボタン)に追加し、スクリーンリーダーはまず元のコンテントを読み、そして [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) のコンテントへを読み取るからです。

Reviewer: ```suggestion
これは describeChild が true にされている場合、Tooltip は [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) を子要素(ボタン)に追加し、スクリーンリーダーはまず元のコンテンツを読み、そして [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) のコンテンツを読み取るからです。
```

> #### Tooltip represents the purpose of the element

Reviewer: ```suggestion
#### Tooltip が要素自体の目的を表す場合
```

Tooltip が要素自体の目的を表す（要素のラベルを提供する）場合は、describeChild を false に設定します。<br/>
例を見てみましょう。<br/>

> この例では、InputタグをテーブルのNameフィールドに設定していますが、Inputタグはスクリーンリーダーで "Name" という単語を読み取ることができません、しかしTooltip を使うことでこの問題を解決することができます。<br/>

Reviewer: ```suggestion
この例では、input タグをテーブルの Name フィールドに設定していますが、input タグはスクリーンリーダーで "Name" という単語を読み取ることができません。しかし Tooltip を使うことでこの問題を解決することができます。<br/>
```

> InputコンポーネントのTooltipを作成、Tooltip のタイトルを "Name" に設定、`describeChild` をfalseに設定して、スクリーンリーダーが何を読み取るかを確認します。<br/>

Reviewer: ```suggestion
input コンポーネントに Tooltip を追加、Tooltip のタイトルを "Name" に設定、describeChild を false に設定して、スクリーンリーダーが何を読み取るかを確認します。<br/>
```

```javascript
const renderName = cellData => {
  const input = document.createElement('input');
  ...
  const tooltipForInput = new Kuc.Tooltip({
    title: 'Name',
    container: input,
    describeChild: false,
    placement: 'bottom'
  });
  return tooltipForInput;
};
```

![render](/img/tooltip_describeChild_false.gif)

> 今回は `describeChild` をfalseに設定した場合、入力にフォーカスが当たったときににスクリーンリーダーは Tooltip のタイトル　"Name" の内容を読み取ります。<br/>

Reviewer: ```suggestion
今回 describeChild を false に設定すると、input にフォーカスが当たったときにスクリーンリーダーは Tooltip のタイトル "Name" の内容を読み取ります。<br/>
```

> これは `describeChild` がfalseの場合、Tooltip は [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) を子要素 (input, fieldGroup) に追加し、スクリーンリーダーは [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) の内容を読み取るからです。<br/>

Reviewer: ```suggestion
これは describeChild が false の場合、Tooltip は [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) を子要素 (input, FieldGroup) に追加し、スクリーンリーダーは [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) の内容を読み取るからです。<br/>
```
